staging/src/k8s.io/client-go/tools/cache/mutation_detector_test.go

```diff
@@ -46,33 +45,30 @@ func TestMutationDetector(t *testing.T) {
 	}
 	stopCh := make(chan struct{})
 	defer close(stopCh)
-	addReceived := make(chan bool)
 	mutationFound := make(chan bool)
 
 	informer := NewSharedInformer(lw, &v1.Pod{}, 1*time.Second).(*sharedIndexInformer)
-	informer.cacheMutationDetector = &defaultCacheMutationDetector{
+	detector := &defaultCacheMutationDetector{
 		name:   "name",
 		period: 1 * time.Second,
 		failureFunc: func(message string) {
 			mutationFound <- true
 		},
 	}
-	informer.AddEventHandler(
-		ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) {
-				addReceived <- true
-			},
-		},
-	)
+	informer.cacheMutationDetector = detector
 	go informer.Run(stopCh)
 
 	fakeWatch.Add(pod)
 
-	select {
-	case <-addReceived:
-	}
+	wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		detector.addedObjsLock.Lock()
+		defer detector.addedObjsLock.Unlock()
+		return len(detector.addedObjs) > 0, nil
+	})
 
+	detector.compareObjectsLock.Lock()
 	pod.Labels["change"] = "true"
+	detector.compareObjectsLock.Unlock()
 
 	select {
 	case <-mutationFound:
```

Comment:
should `mutationFound` have a buffer? Mutation is one line above and I wonder if race may happen here (theoretically) when added item to channel before it is being read.

Reply:
no? this just makes sure an item is added to mutationFound eventually... the detection happens in a background goroutine and isn't blocked

Reply:
what if the message to `mutationFound` will be attempted to be added before this line will start listening. Very low chance I understand

Reply:
Adding the message will block until this line is ready to receive since unbuffered channels are synchronous.
But the message is being added in a different goroutine, and that one being blocked is not an issue ..?

Reply:
> But the message is being added in a different goroutine, and that one being blocked is not an issue

exactly

Reply:
thank you.